vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/DocumentV1ApiHandler.java

```diff
@@ -592,46 +594,60 @@ private static void loggingException(Exceptions.RunnableThrowingIOException runn
 
     // ---------------------------------------------Document Operations ----------------------------------------
 
-    @FunctionalInterface
-    interface Operation {
+    private static abstract class Operation {
+
+        private final Lock lock = new ReentrantLock();
+        private final HttpRequest request;
+        private final ResponseHandler handler;
+        private Supplier<Boolean> operation;
+
+        Operation(HttpRequest request, ResponseHandler handler) {
+            this.request = request;
+            this.handler = handler;
+        }
 
         /**
          * Attempts to dispatch this operation to the document API, and returns whether this completed or not.
          * This return {@code} true if dispatch was successful, or if it failed fatally; or {@code false} if
          * dispatch should be retried at a later time.
          */
-        boolean dispatch();
-
-        /** Wraps the operation parser in an Operation that is parsed the first time it is attempted dispatched. */
-        static Operation lazilyParsed(HttpRequest request, ResponseHandler handler, Supplier<Operation> parser) {
-            AtomicReference<Operation> operation = new AtomicReference<>();
-            return () -> {
-                try {
-                    return operation.updateAndGet(value -> value != null ? value : parser.get()).dispatch();
-                }
-                catch (IllegalArgumentException e) {
-                    badRequest(request, e, handler);
-                }
-                catch (RuntimeException e) {
-                    serverError(request, e, handler);
-                }
+        boolean dispatch() {
+            if (request.isCancelled())
```

Comment:
This is now also run for all operations — not just those against the async-session.